crates/ruff_python_ast/src/identifier.rs

```diff
@@ -0,0 +1,587 @@
+                        .expect("Unable to identify identifier in pattern")
+                }
+            }
+            Pattern::MatchStar(ast::PatternMatchStar { name: Some(_), .. }) => {
+                // Identify `z` in:
+                // ```python
+                // match x:
+                //     case *z:
+                //         ...
+                // ```
+                IdentifierTokenizer::starts_at(self.start(), locator.contents())
+                    .next()
+                    .expect("Unable to identify identifier in pattern")
+            }
+            _ => self.range(),
+        }
+    }
+}
+
+impl Identifier for Excepthandler {
+    /// Return the [`TextRange`] of a named exception in an [`Excepthandler`].
+    ///
+    /// For example, return the range of `e` in:
+    /// ```python
+    /// try:
+    ///     ...
+    /// except ValueError as e:
+    ///     ...
+    /// ```
+    fn identifier(&self, locator: &Locator) -> TextRange {
+        // The exception name is the first identifier token after the `as` keyword.
+        let Excepthandler::ExceptHandler(ast::ExcepthandlerExceptHandler { type_, .. }) = self;
+        IdentifierTokenizer::starts_at(type_.as_ref().unwrap().end(), locator.contents())
+            .nth(1)
+            .expect("Failed to find exception identifier in exception handler")
+    }
+}
+
+/// Return the [`TextRange`] for every name in a [`Stmt`].
+///
+/// Intended to be used for `global` and `nonlocal` statements.
+///
+/// For example, return the ranges of `x` and `y` in:
+/// ```python
+/// global x, y
+/// ```
+pub fn names<'a>(stmt: &Stmt, locator: &'a Locator<'a>) -> impl Iterator<Item = TextRange> + 'a {
+    // Given `global x, y`, the first identifier is `global`, and the remaining identifiers are
+    // the names.
+    IdentifierTokenizer::new(locator.contents(), stmt.range()).skip(1)
+}
+
+/// Return the [`TextRange`] of the `except` token in an [`Excepthandler`].
+pub fn except(handler: &Excepthandler, locator: &Locator) -> TextRange {
+    IdentifierTokenizer::new(locator.contents(), handler.range())
+        .next()
+        .expect("Failed to find `except` token in `Excepthandler`")
+}
+
+/// Return the [`TextRange`] of the `else` token in a `For`, `AsyncFor`, or `While` statement.
+pub fn else_(stmt: &Stmt, locator: &Locator) -> Option<TextRange> {
+    let (Stmt::For(ast::StmtFor { body, orelse, .. })
+        | Stmt::AsyncFor(ast::StmtAsyncFor { body, orelse, .. })
+        | Stmt::While(ast::StmtWhile { body, orelse, .. })) = stmt else {
+        return None;
+    };
+
+    if orelse.is_empty() {
+        return None;
+    }
+
+    IdentifierTokenizer::starts_at(
+        body.last().expect("Expected body to be non-empty").end(),
+        locator.contents(),
+    )
+    .next()
+}
+
+/// Return `true` if the given character is a valid identifier character.
+fn is_python_identifier(c: char) -> bool {
+    c.is_alphanumeric() || c == '_' || c == '.'
```

Comment:
This seems to contradict https://docs.python.org/3/reference/lexical_analysis.html#identifiers, or is this a different kind of identifier?

Reply:
Fixed!